src/Formats/FormatSchemaInfo.cpp

```diff
@@ -34,55 +35,69 @@ FormatSchemaInfo::FormatSchemaInfo(const String & format_schema, const String &
     {
         size_t colon_pos = format_schema.find(':');
-        if ((colon_pos == String::npos) || (colon_pos == 0) || (colon_pos == format_schema.length() - 1)
-            || path.assign(format_schema.substr(0, colon_pos)).makeFile().getFileName().empty())
+        if ((colon_pos == String::npos) || (colon_pos == 0) || (colon_pos == format_schema.length() - 1))
         {
             throw Exception(
                     "Format schema requires the 'format_schema' setting to have the 'schema_file:message_name' format"
                     + (default_file_extension.empty() ? "" : ", e.g. 'schema." + default_file_extension + ":Message'") +
-                    ". Got '" + format_schema
-                    + "'",
-                    ErrorCodes::BAD_ARGUMENTS);
+                    ". Got '" + format_schema + "'", ErrorCodes::BAD_ARGUMENTS);
+        }
+        else
+        {
+            path = fs::path(format_schema.substr(0, colon_pos));
+            String filename = path.has_filename() ? path.filename() : path.parent_path().filename();
+            if (filename.empty())
+                throw Exception(
+                    "Format schema requires the 'format_schema' setting to have the 'schema_file:message_name' format"
+                    + (default_file_extension.empty() ? "" : ", e.g. 'schema." + default_file_extension + ":Message'") +
+                    ". Got '" + format_schema + "'", ErrorCodes::BAD_ARGUMENTS);
         }
-
         message_name = format_schema.substr(colon_pos + 1);
     }
     else
-        path.assign(format_schema).makeFile().getFileName();
+    {
+        path = fs::path(format_schema);
+        if (!path.has_filename())
+            path = path.parent_path() / "";
+    }
 
     auto default_schema_directory = [&format_schema_path]()
     {
-        static const String str = Poco::Path(format_schema_path).makeAbsolute().makeDirectory().toString();
+        static const String str = fs::absolute(format_schema_path) / "";
         return str;
     };
 
-    if (path.getExtension().empty() && !default_file_extension.empty())
-        path.setExtension(default_file_extension);
+    if (!path.has_extension() && !default_file_extension.empty())
+        path = path.parent_path() / (path.stem().string() + '.' + default_file_extension);
 
-    if (path.isAbsolute())
+    if (path.is_absolute())
     {
         if (is_server)
-            throw Exception("Absolute path in the 'format_schema' setting is prohibited: " + path.toString(), ErrorCodes::BAD_ARGUMENTS);
-        schema_path = path.getFileName();
-        schema_directory = path.makeParent().toString();
+            throw Exception("Absolute path in the 'format_schema' setting is prohibited: " + path.string(), ErrorCodes::BAD_ARGUMENTS);
+        schema_path = path.filename();
+        schema_directory = path.parent_path() / "";
     }
-    else if (path.depth() >= 1 && path.directory(0) == "..")
+    else if (path.string().starts_with(".."))
```

Comment:
It looks suspicious and probably we can simplify it

Reply:
Seems like it does not protect from paths like `./dir/../../../some_file`, because both `path.directory(0) == ".."` and `path.string().starts_with("..")` will return false, but path goes outside the format_schema_path directory. It's not related to your changes, but could you please fix it, by the way?